apps/allocations/app/hooks/usePathHelpers.js

```diff
@@ -0,0 +1,47 @@
+import React from 'react'
+import { usePath } from '../api-react'
+
+export default function usePathHelpers() {
+  const [ path, requestPath ] = usePath()
+  const [ firstTry, setFirstTry ] = React.useState(true)
+
+  // accepts a pattern like '/budgets/:id', where ':id' is a named parameter
+  // redirects to '/' if the current path doesn't match at all
+  // otherwise, returns an object with keys matching the named parameters and
+  // values filled in from the current path
+  const parsePath = React.useCallback(pattern => {
+    const namedParameters = pattern.match(/(:[a-zA-Z]+)/g)
+
+    // replace named paramaters with regex-compatible "named groups"
+    namedParameters.forEach(x => {
+      pattern = pattern.replace(x, `(?<${x.slice(1)}>[a-zA-Z0-9=-]+)`)
+    })
+
+    const matchData = path.match(pattern)
+    if (!matchData) {
+      requestPath('/')
+      return {}
+    }
+
+    return matchData.groups
+  }, [ path, requestPath ])
+
+  // if this page is the first page loaded for the user,
+  // we may still have incomplete state from aragonAPI;
+  // let's give it a second before redirecting
+  const patientlyRequestPath = React.useCallback(path => {
+    if (firstTry) {
+      setTimeout(() => setFirstTry(false), 1000)
+    } else {
+      requestPath(path)
+    }
+  }, [ path, requestPath ])
```

Comment:
I cannot get this to work properly. If I put the wrong ID into the url, I'm left with a blank page. If I log `firstTry` it always returns `true`. I think `usePathHelpers` must be getting reloaded and the state keeps getting reset.

Reply:
Ah, thanks! This broke when I started using `useCallback` with the definition of this function. We need to redefine the function when the `firstTry` variable changes. Fixed locally; pushing soon.